### PR TITLE
Simple FOV

### PR DESCRIPTION
**Pull request by nightblade9**
- Basic FOV calculation (distance between points)
- Expensive calculation is cached, not run each frame
- Monsters only attack if you're in their FOV
- Basic (hideous) fog-of-war for walls

### CODE REVIEW

File: src/Scripts/Ecs/Core/Entity.gd [hunk 1]
```diff
@@ -4,6 +4,7 @@ const SpriteComponent = preload("res://Scripts/Ecs/Components/SpriteComponent.gd
 
 # type (as a string) => instance
 var _components = {}
+var sight_radius = 0
```
Comment (3akev) [inline src/Scripts/Ecs/Core/Entity.gd]: Shouldn't this be in a component? Walls don't have sight, do they?

File: src/Scripts/Ecs/Systems/DisplaySystem.gd [hunk 2]
```diff
@@ -4,24 +4,52 @@ var _tiles_tilemap:TileMap
...
+
+func _is_in_player_fov(entity):
+	if entity.position.x < _player.position.x - _player.sight_radius or \
+		entity.position.y < _player.position.y - _player.sight_radius or \
+		entity.position.x > _player.position.x + _player.sight_radius or \
+		entity.position.y > _player.position.y + _player.sight_radius:
+			
+		return false
+		
+	return sqrt(pow(entity.position.x - _player.position.x, 2) + pow(entity.position.y - _player.position.y, 2)) <= _player.sight_radius
```
Comment (3akev) [inline src/Scripts/Ecs/Systems/DisplaySystem.gd]: Use `entity.position.distance_to(_player.position)`. Built in methods are in C++, therefore probably faster, so we can get away with dropping that unsightly AABB too. Is it really necessary? Seems like premature optimization to me.

File: src/Scripts/Ecs/Systems/ChasePlayerSystem.gd [hunk 3]
```diff
@@ -18,11 +18,17 @@ func on_turn_end():
...
 	for entity in entities:
-		var direction = entity.position.direction_to(_player.position)
-		if abs(direction.x) > abs(direction.y):
-			_event_bus.emit_signal("move_entity", entity, entity.position.x + sign(direction.x), entity.position.y)
-		else:
-			_event_bus.emit_signal("move_entity", entity, entity.position.x, entity.position.y + sign(direction.y))
+		# Move only if player is in monster FOV.
+		# Expensive but accurate, assumes few monsters and done once per turn.
+		# If this is too expensive, switch to manhattan distance (no sqrt)
+		var distance = sqrt(pow(entity.position.x - _player.position.x, 2) + pow(entity.position.y - _player.position.y, 2))
```
Comment (3akev) [inline src/Scripts/Ecs/Systems/ChasePlayerSystem.gd]: Pass in `entity.position` and use `position.distance_to(_player.position)`.

File: src/Scripts/Ecs/Systems/DisplaySystem.gd [hunk 4]
```diff
@@ -4,24 +4,74 @@ var _tiles_tilemap:TileMap
...
 		var component = entity.get("SpriteComponent")
-
 		var tilemap = tilemaps_by_name[component.layer]
-		var tile_index = tilemap.tile_set.find_tile_by_name(component.tile_name)
-		tilemap.set_cell(entity.position.x, entity.position.y, tile_index)
\ No newline at end of file
+		var key = "%s, %s" % [entity.position.x, entity.position.y]
+		
+		# It's in our FOV, or it's a wall in fog-of-war
+		if  (_fov_cache.has(key) and _fov_cache[key] == true) or \
```
Comment (3akev) [inline src/Scripts/Ecs/Systems/DisplaySystem.gd]: Can be simplified to `_fov_cache.get(key) == true`. `.get()` returns null if not found.

File: src/Scripts/Ecs/Systems/DisplaySystem.gd [hunk 5]
```diff
@@ -4,24 +4,74 @@ var _tiles_tilemap:TileMap
...
+
+func _is_in_player_fov(x, y):
+	if x < _player.position.x - _player.sight_radius or \
+		y < _player.position.y - _player.sight_radius or \
+		x > _player.position.x + _player.sight_radius or \
+		y > _player.position.y + _player.sight_radius:
+			
+		return false
+		
+	return sqrt(pow(x - _player.position.x, 2) + pow(y - _player.position.y, 2)) <= _player.sight_radius
```
Comment (3akev) [inline src/Scripts/Ecs/Systems/DisplaySystem.gd]: Ditto. Use built-in methods, please.

File: src/Scripts/Ecs/Core/Entity.gd (hunk 1 above)
Comment (nightblade9) [reply src/Scripts/Ecs/Core/Entity.gd]: It's true that walls don't have sight. Which component would sight go into? ... I guess `SightComponent` 👀

File: src/Scripts/Ecs/Systems/DisplaySystem.gd (hunk 2 above)
Comment (nightblade9) [reply src/Scripts/Ecs/Systems/DisplaySystem.gd]: AABB is very fast and computationally cheap. It saves us from expensive distance calculations. You're right that `distance_to` is probably better - will do.

**Pull request merged**